sematic/ui/src/pipelines/PipelinePanels.tsx

```diff
@@ -1,55 +1,88 @@
-import RunPanel from "./RunPanel";
 import { ExtractContextType } from "../components/utils/typings";
-import { usePipelineRunContext } from "../hooks/pipelineHooks";
+import { useGraph } from "../hooks/graphHooks";
+import { selectedRunHashAtom, usePipelineRunContext } from "../hooks/pipelineHooks";
 import { Run } from "../Models";
+import GraphContext from "./graph/graphContext";
+import MenuPanel from "./MenuPanel";
+import NotesPanel from "./NotesPanel";
 import PipelinePanelsContext from "./PipelinePanelsContext";
 import PipelineRunViewContext from "./PipelineRunViewContext";
-import GraphContext from "./graph/graphContext";
-import { useGraph } from "../hooks/graphHooks";
+import RunPanel from "./RunPanel";
+
+const sessionStorageKey = 'previous_run_path';
 
 export default function PipelinePanels() {
   const { rootRun }
     = usePipelineRunContext() as ExtractContextType<typeof PipelineRunViewContext> 
     & {
       rootRun: Run
     };
-
+    
+  const [selectedRunId, setSelectedRunId] = useAtom(selectedRunHashAtom);
   const [selectedPanelItem, setSelectedPanelItem] = useState("run");
-  const [selectedRun, setSelectedRun] = useState<Run>(rootRun);
-  const pipelinePanelsContext = useMemo<ExtractContextType<typeof PipelinePanelsContext>>(() => ({
-    selectedPanelItem, setSelectedPanelItem,
-    selectedRun, setSelectedRun
-  }), [selectedPanelItem, setSelectedPanelItem, selectedRun, setSelectedRun]);
 
   const [graph, isGraphLoading, error] = useGraph(rootRun.id);
   const graphContext = useMemo<ExtractContextType<typeof GraphContext>>(() => ({
     graph,
     isLoading: isGraphLoading
   }), [graph, isGraphLoading]);
 
-  // Update the selectedRun with the data coming from the graph
-  useEffect(() => {
+  const isPreviousStateRecovered = useRef(false);
+
+  const selectedRun = useMemo(() => {
+    let runId = selectedRunId;
+    if (!runId) {
+      runId = rootRun.id;
+    }
     if (!graph) {
-      return;
+      return undefined;
     }
-    if (selectedRun && graph.runs) {
-      let newSelectedRun = graph.runsById.get(selectedRun.id);
-      if (!newSelectedRun) {
-        newSelectedRun = rootRun;
-        graph.runs.forEach((run: Run) => {
-          if (run.calculator_path === selectedRun.calculator_path) {
-            newSelectedRun = run;
-          }
-        });
+    return graph.runsById.get(runId) || rootRun;
+  }, [selectedRunId, graph, rootRun]);
+
+  const pipelinePanelsContext = useMemo<ExtractContextType<typeof PipelinePanelsContext>>(() => ({
+    selectedPanelItem, setSelectedPanelItem,
+    selectedRun, setSelectedRunId
+  }), [selectedPanelItem, setSelectedPanelItem, selectedRun, setSelectedRunId]);
+
+  useEffect(()=> {
+    // In case of fresh mounting, highlighten the previously selected run 
+    // on a best effort. However, in case of multiple runs share the same
+    // name, we might not necessarily highlighten the right one.
+    if (!!graph && isPreviousStateRecovered.current === false) {
```

Comment:
Hi @neutralino1, this code is to keep the logic of deleted line from L40-L48, to `have the same nested run selected when users change root run with the drop down. ` 

Since nested runs might have the same run path, we might only sometimes be able to select the right nest run as previously selected. This flaky feature might not be worth the benefits it gains. Shall we remove it?

Reply:
Thanks for confirming. It is removed now.